### PR TITLE
Add --pep420-mode to deal with PEP 420

### PR DESCRIPTION
**Pull request by estan**
A note at [Good Integration Practices](http://doc.pytest.org/en/latest/goodpractices.html) has the following to say about PEP 420:

```text
You can use Python3 namespace packages (PEP420) for your application but pytest
will still perform test package name discovery based on the presence of __init__.py
files. If you use one of the two recommended file system layouts above but leave away
the __init__.py files from your directories it should just work on Python3.3 and above.
From “inlined tests”, however, you will need to use absolute imports for getting at
your application code.
```

But contrary to this, the presence of PEP-420 namespace packages is problematic.

Consider this project `proj`, which provides a package `pkg` in the PEP 420 namespace package `suite`:

```
proj
proj/suite
proj/suite/pkg
proj/suite/pkg/tests
proj/suite/pkg/tests/test_foo.py
proj/suite/pkg/tests/__init__.py
proj/suite/pkg/foo.py
proj/suite/pkg/__init__.py
proj/setup.py
```

**foo.py**
```python
def func():
    return 'foo'
```

**test_foo.py**
```python
import sys

from suite.pkg.foo import func

def test_func():
    assert func() == 'foo'
```

A simple `py.test` in the project directory will fail, as expected, during the import of `test_foo.py` with `ImportError: No module named 'suite'`, because `pytest` will prepend `/path/to/proj/suite` to `sys.path`, not `/path/to/proj` (the upwards traversal when determining the test module name, and the path to prepend, stopped at `suite`).

The simple solution is to manually add `/path/to/proj` to `PYTHONPATH`. This "fixes" the problem (the `import` in the test will now work), but is less than ideal, because having `/path/to/proj/suite` in `sys.path` (which is still there as well, and needs to be there since `pytest` will import `test_foo.py` by the name `pkg.tests.test_foo`) could lead to nasty surprises.

This pull request adds a `--pep420-mode`, in which the traversal when determining the module name is not based on the presence of `__init__.py`, but instead stops when it hits a directory that is in `sys.path`. The mode is intended for when you have PEP 420 namespace packages, and when using this mode it is your responsibility to ensure `sys.path` has the appropriate path(s) (typically your project root).

It's not meant to be merged as-is, but is meant to start a discussion about how `py.test` should handle the presence of PEP 420 packages. What do you think, could a mode like this be a good solution?

I don't think `pytest` can do correct "file system path" -> "dotted module name" deduction in the presence of PEP 420 packages without some form of additional user input (such as configuring `PYTHONPATH` properly). I know `setuptools` is also struggling with PEP 420 in its [issue 97](https://github.com/pypa/setuptools/issues/97).

### CODE REVIEW

**Comment by estan** [comment]
If this solution is OK in principle, I guess the right thing to do would be to give `LocalPath.pyimport` (in `py.path`) a new `isinpath=False` parameter. When `True`, the method would not modify `sys.path` (i.e. ignoring the `ensuresyspath` parameter), and would use the traversal behavior I suggest here instead of looking for `__init__.py`.

Then here in `pytest` I think it would be simply:

```python
            if self.config.getoption("--pep420-mode"):
                mod = self.fspath.pyimport(isinpath=True)
            else:
                mod = self.fspath.pyimport(ensuresyspath=importmode)
```

**Comment by jaraco** [comment]
It strikes me as the wrong approach to have an opt-in flag for PEP 420 packages, as one can imagine a future where this behavior is the default, standard practice. I would reframe the question - how should pytest behave in a PEP 420 world, where any folder is potentially a package? Then implement that as the default behavior, and consider backward-compatibility support.

**Comment by estan** [comment]
@jaraco Good point, and I agree (but don't stare too much at the name `--pep420-mode`, that was just a name I gave it in lack of a better idea).

But do you agree that in the face of PEP 420 packages, it's impossible to deduce the module name from the module file path without further input from the user to guide it? I've been trying to come up with an idea, but I don't think it's possible.

We could simply make the behavior that my flag introduces the default, and instead add a `--compat-something` flag to enable the old behavior. But would that be acceptable at this point? I think it would break things for a lot of people who currently assume they can run `py.test` without having to add their project path to `PYTHONPATH`.

I don't think it's possible to have a single mode that supports both the convenient behavior that non-PEP 420-using projects are used to right now, and at the same time works correctly for PEP 420-using projects :(

**Comment by estan** [comment]
Either something like that, or we introduce a `namespace_packages` parameter to `find_packages` and require users to do

```python
namespace_packages = [...]
setup(
    ...
    packages=find_packages(namespace_packages=namespace_packages),
    namespace_packages=namespace_packages,
    ...
)
```

to avoid having to repeat the list of namespace packages literally. Not sure I like that..

Yet another possibility is that the `namespace_packages` parameter to `find_packages` could be the sentient value `IMPLICIT` (obviously needs better name), like:

```python
setup(
    ...
    packages=find_packages(namespace_packages=IMPLICIT),
    namespace_packages=namespace_packages,
    ...
)
```

Where, if `IMPLICIT` is given, `find_packages` would enter the "noop" mode I described earlier, and the actual search would be deferred until `namespace_packages` (the `setup` parameter) is known. Not sure I like that either. Would take some careful wording in the doc so that users understand that `IMPLICIT` does not make sense if they're using `find_packages` in some other context (not the `setup` call). Maybe it would be better to trigger the deferred behavior with a new `bool` parameter `consider_namespace_packages` (would also require careful wording in the docs).

**Comment by estan** [comment]
Though thinking about it, I am warming up to the idea of

```python
namespace_packages = [...]
setup(
    ...
    packages=find_packages(namespace_packages=namespace_packages),
    namespace_packages=namespace_packages,
    ...
)
```

Because, we've now discovered that to do its job properly, `find_packages` needs more info. It then only make sense that `find_packages` takes that info as a parameter. It seems the simplest solution. Personally I could live with having to declare a variable to avoid repeating the list of namespace packages.

**Comment by estan** [comment]
The parameter could of course be called `implicit_namespace_packages`, as it is only those that `find_packages` strictly need, even if it's a bit long.

**Comment by estan** [comment]
Or (sorry for all the ideas here), we could change the meaning of the `include` parameter, such that a directory in `include`is considered a package even if it lacks an `__init__.py`, and then instruct users to add their implicit namespace packages to `include`. I think I like this idea the best.

**Comment by estan** [comment]
Oh oh oh! I'm sooo sorry. My last five (!) comments were meant for [setuptools issue #97](https://github.com/pypa/setuptools/issues/97) ! Will delete them and move it there.

**Comment by nicoddemus** [comment]
Hey @estan first of all thanks for the time you put on writing the PR, we appreciate it (and sorry for taking so long to respond).

Regarding PEP420 support, I think perhaps we should first see what comes out pypa/setuptools#97 before attempting to support it in pytest. I would not like to end up with two different approaches (say `--pep420-mode` for pytest and `__nsdir__` files for setuptools), that would just add to the overall confusion. 

It is really a complicated issue at the moment, and I feel this should be supported directly by Python instead of having each build/test tool implement their own "filepath -> module mapping".

In conclusion, do you agree we should perhaps close this PR as it stands?

**Comment by estan** [comment]
@nicoddemus: Right, I've been somewhat active on https://github.com/pypa/setuptools/issues/97 as well, and I guess it's better to see how that falls out (thought it's been open for a long time..).

When it comes to Python supporting a "filepath -> module mapping" on its own, while I think it could happen, I'm just afraid it's going to take a long time, and surely py.test should work well on versions that don't have this capability anyway?

But yes, I'll close this until setuptools #97 is settled.
